python/triton/compiler.py

```diff
@@ -1384,58 +1430,42 @@ def compile(fn, **kwargs):
-        metadata = {"num_warps": num_warps, "num_stages": num_stages, "ctime": dict()}
-    # build compilation stages
-    stages = {
-        "ast": (lambda path: fn, None),
-        "ttir": (lambda path: _triton.ir.parse_mlir_module(path, context),
-                 lambda src: ast_to_ttir(src, signature, configs[0], constants)),
-        "ttgir": (lambda path: _triton.ir.parse_mlir_module(path, context),
-                  lambda src: ttir_to_ttgir(src, num_warps, num_stages, compute_capability)),
-        "llir": (lambda path: Path(path).read_bytes(),
-                 lambda src: ttgir_to_llir(src, extern_libs, compute_capability)),
-        "ptx": (lambda path: Path(path).read_text(),
-                lambda src: llir_to_ptx(src, compute_capability)),
-        "cubin": (lambda path: Path(path).read_bytes(),
-                  lambda src: ptx_to_cubin(src, compute_capability))
-    }
+      metadata = {"num_warps": num_warps, "num_stages": num_stages, "ctime": dict()}
+      if ext == "ptx":
+        assert "shared" in kwargs, "ptx compilation must provide shared memory size"
+        metadata["shared"] = kwargs["shared"]
+
     first_stage = list(stages.keys()).index(ext)
     asm = dict()
     module = fn
     # run compilation pipeline  and populate metadata
     for ir, (parse, compile) in list(stages.items())[first_stage:]:
-        path = fn_cache_manager._make_path(f"{name}.{ir}")
-        if ir == ext:
-            next_module = parse(fn)
-        elif os.path.exists(path) and \
-                ir in metadata["ctime"] and \
-                os.path.getctime(path) == metadata["ctime"][ir]:
-            next_module = parse(path)
-        else:
-            next_module = compile(module)
-            fn_cache_manager.put(next_module, f"{name}.{ir}")
-        if os.path.exists(path):
-            metadata["ctime"][ir] = os.path.getctime(path)
-        asm[ir] = next_module if ir == "cubin" else str(next_module)
-        if ir == "llir" and "shared" not in metadata:
-            metadata["shared"] = _triton.get_shared_memory_size(module)
-        if ir == "ptx":
-            metadata["name"] = ptx_get_kernel_name(next_module)
-        module = next_module
+      path = fn_cache_manager._make_path(f"{name}.{ir}")
+      if ir == ext:
+        next_module = parse(fn)
+      elif os.path.exists(path) and\
+           ir in metadata["ctime"] and\
+           os.path.getctime(path) == metadata["ctime"][ir]:
+        next_module = parse(path)
+      else:
+        next_module = compile(module)
+        fn_cache_manager.put(next_module, f"{name}.{ir}")
+      if os.path.exists(path):
+        metadata["ctime"][ir] = os.path.getctime(path)
+      asm[ir] = next_module if ir == "cubin" else str(next_module)
+      if ir == "llir" and "shared" not in metadata:
+        metadata["shared"] = _triton.get_shared_memory_size(module)
+      if ir == "ptx":
+        metadata["name"] = ptx_get_kernel_name(next_module)
+      module = next_module
```

Comment:
We might need a universal auto-format for python later, noticing the indent is not identical to other methods in compiler.py.

Reply:
Yep. I found `black` too aggressive in the past -- especially for triton.ops and tutorials where kernels have 50 arguments, but we could add #noformat directives. Not a big priority though right now

Reply:
Yeah.